build(aio): add support for using the locally built Angular packages for aio

gkalpak opened the pull request:
This commit allows building angular.io against the locally built Angular packages. It adds two new npm scripts:

- `setup-local`: Same as `setup`, but overwrites the Angular packages for both angular.io and the examples boilerplate with the locally built ones.
- `build-local`: Same as `build`, but uses `setup-local` instead of `setup` under the hood, thus overwriting installed Angular packages with locally built ones.

Fixes #18611.

gkalpak committed:
build(aio): add support for using the locally built Angular packages for aio

This commit allows building angular.io against the locally built Angular
packages. It adds two new npm scripts:

- `setup-local`: Same as `setup`, but overwrites the Angular packages for both
  angular.io and the examples boilerplate with the locally built ones.
- `build-local`: Same as `build`, but uses `setup-local` instead of `setup`
  under the hood, thus overwriting installed Angular packages with locally built
  ones.

Fixes #18611

petebacondarwin commented:
Various failures on Travis.

mary-poppins commented:
You can preview 0742e1b at https://pr18722-0742e1b.ngbuilds.io/.

gkalpak commented on scripts/ci/test-aio.sh:
```diff
@@ -18,12 +18,6 @@ source ${thisDir}/_travis-fold.sh
 
-  # Run unit tests for boilerplate tools
-  travisFoldStart "test.aio.boilerplate.unit"
-    yarn boilerplate:test
```
This is now part of `yarn tools-test`:

https://github.com/gkalpak/angular/blob/f739b94fcaef280487377fbd6ef87f4c20b36f91/aio/package.json#L39

...which is run in `test-aio-tools.sh`:

https://github.com/gkalpak/angular/blob/f739b94fcaef280487377fbd6ef87f4c20b36f91/scripts/ci/test-aio-tools.sh#L13

mary-poppins commented:
You can preview 90a1837 at https://pr18722-90a1837.ngbuilds.io/.

mary-poppins commented:
You can preview f99b8bb at https://pr18722-f99b8bb.ngbuilds.io/.

petebacondarwin commented on aio/package.json:
```diff
@@ -8,14 +8,20 @@
     "e2e": "ng e2e --no-webdriver-update",
-    "setup": "yarn && yarn build-ie-polyfills && yarn boilerplate:remove && yarn boilerplate:add && yarn generate-plunkers && yarn generate-zips && yarn docs",
+    "presetup": "yarn ~~setup-phase1",
+    "setup": "node tools/ng-packages-installer restore . && yarn boilerplate:add",
```
Um, doesn't `yarn boilerplate:add` effectively run `ng-packages-installer restore` already?
Similarly `yarn boilerplate:add --local` effectively runs `ng-packages-installer overwrite`?

gkalpak replied:
`yarn boilerplate:xyz` runs `ng-packages-installer` for the boilerplate files. `yarn setup-local/build-local` run it for `aio/`.

petebacondarwin reviewed (changes requested):
I may be wrong but I think that the scripts `setup` and `setup-local` are calling `ng-packages-installer` unnecessarily. Isn't it enough to do:

```json
"setup": "yarn boilerplate:add",
"setup-local": "yarn boilerplate:add -- --local",
```

gkalpak commented on aio/package.json:
```diff
@@ -41,7 +46,12 @@
-    "build-ie-polyfills": "node node_modules/webpack/bin/webpack.js -p src/ie-polyfills.js src/generated/ie-polyfills.min.js"
+    "build-ie-polyfills": "node node_modules/webpack/bin/webpack.js -p src/ie-polyfills.js src/generated/ie-polyfills.min.js",
+    "~~build": "ng build --target=production --environment=stable -sm --build-optimizer",
+    "post~~build": "yarn sw-manifest && yarn sw-copy",
```
Not really. It was like this before (as `build` and `postbuild`), so I just added the `~~` prefix.

gkalpak commented on aio/package.json:
```diff
@@ -8,14 +8,20 @@
+    "postsetup": "yarn ~~setup-phase3",
+    "presetup-local": "yarn ~~setup-phase1",
+    "setup-local": "node tools/ng-packages-installer overwrite . && yarn boilerplate:add -- --local",
+    "postsetup-local": "yarn ~~setup-phase3",
```
The initial idea for not having `"presetup-local": "yarn presetup"` was that it is not obvious (if you don't look carefully through the scripts) that `"presetup": "..."` applies to `setup-local` too. So, someone might make a change that is specific to `setup` not realizing that it will affect `setup-local` too.

Given that the likelihood of this happening is very low, I am happy to reduce the number of scripts :smiley:

gkalpak commented:
The extra calls are for `aio/` itself (not boilerplate).
Fixed the other comments.

mary-poppins commented:
You can preview 551c5ad at https://pr18722-551c5ad.ngbuilds.io/.

petebacondarwin commented:
We agreed to change this so that you can switch from being in "npm" mode to "local" mode and back; and then all the other commands work in the current mode, e.g. build, serve, etc.

mary-poppins commented:
You can preview 1a4ae90 at https://pr18722-1a4ae90.ngbuilds.io/.

mary-poppins commented:
You can preview 07de92a at https://pr18722-07de92a.ngbuilds.io/.

mary-poppins commented:
You can preview 0d0853e at https://pr18722-0d0853e.ngbuilds.io/.

mary-poppins commented:
You can preview 02a4e23 at https://pr18722-02a4e23.ngbuilds.io/.

mary-poppins commented:
You can preview ed16645 at https://pr18722-ed16645.ngbuilds.io/.

mary-poppins commented:
You can preview a157d0e at https://pr18722-a157d0e.ngbuilds.io/.

gkalpak commented:
I played around with decoupling npm/local packages from `yarn setup` and run into some issues/concerns:

- In `setup`, boilerplate is removed and re-added every time (including `node_modules/`), so it is not straight forward to retain the setting (npm or local) for boilerplate.
- Whenever a dependency is updated (e.g. you checkout an update `yarn.lock`) and `yarn [install]` is run (which happens implicitly for `yarn setup` and the scripts that directly or indirectly depend on it), then the packages are restored (i.e. the ones from npm replace the local ones).

These make it difficult to guarantee that switching to local packages will not be unexpectedly switched under the hood.

We could have mechanisms to ensure that, but they have their cons to. E.g.:

- Using `yarn link` is a little too magical and affects global state. It is for example not possible to have two different local clones (e.g. one for master, one for stable) and independently test each against their respective local packages).
- Somehow cache the current state of Angular packages (i.e. npm or local) and choosing a behavior based on that is unnecessarily complicated.

##
With the current implementation:

- One can use `setup`/`setup-local` (or `build`/`build-local`) as the switch. The former ensures that the packages come form npm (and affects subsequent commands, such as `yarn test`, `yarn e2e` etc). The latter ensures that the local package are used (and similarly affects subsequent commands).
- A warning is printed before any script if there are any local packages used.
- One could still manually switch to npm/local packages (for a specific directory) with `node tools/ng-packages-installer (restore|overwrite) <root/dir>`.

I slightly prefer this approach (with `setup`/`setup-local`), because it is more explicit (less magical), predictable (if you stick to using the scripts, you won't be unexpectedly switched to npm/local packages) and unifies the behavior across the project (e.g. you don't have to switch `boilerplate` separately.

It is not ideal, but maybe good enough for now. @petebacondarwin, wdyt?

petebacondarwin commented:
I'll take a look when I am done with Jesus' PR

petebacondarwin commented:
I think this approach will have to do for now. I haven't had time to go through the PR in detail. Will do so tomorrow.

petebacondarwin commented on aio/package.json:
```diff
@@ -8,31 +8,38 @@
+    "setup": "node tools/ng-packages-installer restore . && yarn boilerplate:add",
+    "postsetup": "yarn build-ie-polyfills && yarn generate-plunkers && yarn generate-zips && yarn docs",
+    "presetup-local": "yarn presetup",
+    "setup-local": "node tools/ng-packages-installer overwrite . && yarn boilerplate:add -- --local",
```
no that we are on yarn v1, we don't need all these `--` to separate the arguments to be passed through.

petebacondarwin commented on aio/package.json:
```diff
@@ -8,31 +8,38 @@
-    "check-env": "node scripts/check-environment",
+    "check-ng-packages": "node tools/ng-packages-installer check",
+    "check-env": "yarn ~~check-env",
+    "postcheck-env": "yarn check-ng-packages -- .",
```
+1!

petebacondarwin commented on aio/tools/examples/example-boilerplate.spec.js:
```diff
@@ -1,57 +1,62 @@
+const shelljs = require('shelljs');
+
+const ngPackagesInstaller = require('../ng-packages-installer');
+const exampleBoilerPlate = require('./example-boilerplate');
```
Ordering alphabetically by const (`ngPackagesInstaller`>`exampleBoilerPlate`!) or by module name (`glob`>`canonical-path`!)?

gkalpak replied:
Or by path (`./<...>` > `../<...>`) :grin:

petebacondarwin replied:
By path (`glob`>`canonical-path`!) :-P

gkalpak replied:
Sigh! 😞

petebacondarwin commented on aio/tools/examples/example-boilerplate.spec.js:
```diff
@@ -81,27 +89,33 @@ describe('example-boilerplate tool', () => {
-      spyOn(fs, 'copySync');
-    });
+    it('should overwrite the Angular packages if `useLocal` is true', () => {
+      ngPackagesInstaller.overwritePackages.and.callFake(() => expect(shelljs.exec).toHaveBeenCalled());
```
This seems a slightly convoluted way of checking that `shelljs.exec` has been called. Is there some reason why you wouldn't just check directly after the following line?

gkalpak replied:
I want to verify the order :grin:
Latest jasmine has some matchers for that (e.g. `toHaveBeenCalledAfter()`), but we're not there yet.

petebacondarwin replied:
Right I see. You could also use a log for that.

petebacondarwin replied:
Wot? You don't like `expect(log).toEqual(['first-thing', 'second-thing', 'third-thing`])` approach to checking the order that spies are called?

gkalpak replied:
I like that particular line :smiley: I don't like the boilerplate to get there :smiley::

```ts
const log = [];
spy1.and.callFake(() => log.push('spy1'));
spy2.and.callFake(() => log.push('spy2'));
expect(log).toEqual(['spy1', 'spy2']);
expect(spy2).toHaveBeenCalledWith('foo');

// vs

spy2.and.callFake(() => expect(spy1).toHaveBeenCalled());
expect(spy2).toHaveBeenCalledWith('foo');
```

petebacondarwin commented on aio/tools/ng-packages-installer.js:
```diff
@@ -0,0 +1,237 @@
+// Constants
+const ROOT_DIR = path.resolve(__dirname, '../..');
+const PACKAGES_DIR = path.join(ROOT_DIR, 'packages');
+const PACKAGES_DIST_DIR = path.join(ROOT_DIR, 'dist/packages-dist');
```
IMO `resolve` is always a safer bet that `join`, just in case the base path is not absolute (which I admit it is in this case, for now)

gkalpak replied:
`resolve` has its issues as well. If you accidentally pass a `/`-prefixed segment, it will ignore all previous segments. So, neither is ideal, but at least there are tests to catch mishaps :smiley:

(Since we are using `canonical-path`, I am actually tempted to change them to `${ROOT_DIR}/packages` and `${ROOT_DIR}/packages-dist` :smiley:)

petebacondarwin replied:
> If you accidentally pass a /-prefixed segment, it will ignore all previous segments
More commonly you actually mean to pass a previously resolved path as a later argument, which then gets corrupted by `join`:

```
const myPath1 = path.join(myBase, getPossiblyAbsolutePathToResource());
const myPath2 = path.resolve(myBase, getPossiblyAbsolutePathToResource());

> myPath1 -> /a/b/c/x/y/z
> myPath2 -> /x/y/z
```

gkalpak replied:
> More commonly you actually mean to pass a previously resolved path as a later argument

I never mean to pass previously resolved paths when using `join()` 😛

petebacondarwin commented on aio/tools/ng-packages-installer.js:
```diff
@@ -0,0 +1,237 @@
+
+    /**
+     * A sorted list of Angular package names.
+     * (Detected as directories in '/packages/' that contain a top-level 'package.json' file.)
```
+100 - we should roll this out across the project :-)

petebacondarwin commented on aio/tools/ng-packages-installer.js:
```diff
@@ -0,0 +1,237 @@
+  _isLocalPackage(packageName, nodeModulesDir) {
+    const targetPackageDir = path.join(nodeModulesDir, '@angular', packageName);
+    const localFlagFile = path.join(targetPackageDir, NG_LOCAL_FILENAME);
+    const isLocal = fs.existsSync(localFlagFile);
```
So if the node_modules package has been overwritten in any other way than via this tool, we would not be able to tell. I guess this is a reasonable constraint.

gkalpak replied:
Yeah, it won't be detected. I think it is a reasonable constraint too.

petebacondarwin commented on aio/tools/ng-packages-installer.js:
```diff
@@ -0,0 +1,237 @@
+    this._log(`Overwriting package '${packageName}' (${sourcePackageDir} --> ${targetPackageDir})...`);
+
+    if (fs.existsSync(targetPackageDir)) {
+      shelljs.rm('-rf', targetPackageDir);
```
If for some reason (in the future) the local package relied upon a different version of a dependency that the rest of the setup, then this folder would contain its own `node_modules` folder, containing the conflicted versions.  By deleting the whole folder we are removing these dependencies.

Are we happy with this limitation?

gkalpak replied:
I am not thrilled, but I can live with it.

Generally, I think we will have the same versions of dependencies as the `@angular` packages in our `yarn.lock`, so I don't expect there to be node_modules dirs inside them. What _can_ happen though is the local packages having newer peerDependencies versions than the ones we have currently installed, which might be a problem. But I don't see that happening often and given that I can't think of a good solution I am happy.

petebacondarwin commented on aio/tools/ng-packages-installer.js:
```diff
@@ -0,0 +1,237 @@
+    const installCmd = 'yarn install --check-files';
+
+    this._log(`Running '${installCmd}' in '${rootDir}'...`);
+    shelljs.exec(installCmd, {cwd: rootDir});
```
I assume `--check-files` checks all the files and not just the `package.json`?

gkalpak replied:
I checked and it works for our usecase (probably doing an integrity check or something). But I admit there is no guarantee it will not break in a future version :grin:

petebacondarwin commented on aio/tools/ng-packages-installer.js:
```diff
@@ -0,0 +1,237 @@
+      this.debug = argv.debug;
+
+      const packagePrefix = '  - @angular/';
+      this._log(`Available Angular packages:\n${packagePrefix}${this.ngPackages.join(`\n${packagePrefix}`)}`);
```
This would render strangely if there were no packages in the array. How about:

```typescript
this._log(`Available Angular packages: ${this.ngPackages.map(p => `\n${packagePrefix}${p}`)}`);
```

gkalpak replied:
The fact that there are no packages would be much stranger than the message 😛
But point taken :smiley:

petebacondarwin commented on aio/tools/ng-packages-installer.spec.js:
```diff
@@ -0,0 +1,282 @@
+
+        const warnings = console.warn.calls.allArgs().map(args => args[0]);
+        expect(warnings[0]).toMatch(restoreCmdRe1);
+        expect(warnings[1]).toMatch(restoreCmdRe2);
```
Any reason not to interleave these expectations with their calls, i.e.:

```
const dir1 = rootDir;
const restoreCmdRe1 = RegExp(`\\bnode .*?ng-packages-installer restore .*?${path.normalize(rootDir)}\\b`);
installer.checkPackages(dir1);
const warnings = console.warn.calls.allArgs().map(args => args[0]);
expect(warnings[0]).toMatch(restoreCmdRe1);
console.warn.reset();

const dir2 = '.';
const restoreCmdRe2 = RegExp('\\bnode .*?ng-packages-installer restore \\.');
installer.checkPackages(dir2);
const warnings = console.warn.calls.allArgs().map(args => args[0]);
expect(warnings[0]).toMatch(restoreCmdRe2);
```

gkalpak replied:
I find it easier to compare, but happy to move them around :smiley:

petebacondarwin commented on aio/tools/ng-packages-installer.spec.js:
```diff
@@ -0,0 +1,282 @@
+      });
+
+      installer.restorePackages(rootDir);
+      expect(installer._findLocalPackages).toHaveBeenCalledWith(absoluteRootDir);
```
```
expect(installer._reinstallOverwrittenNodeModules).not.toHaveBeenCalledWith(absoluteRootDir);
```
here too

petebacondarwin replied:
Oh, I see you test that below.

petebacondarwin commented on aio/tools/ng-packages-installer.spec.js:
```diff
@@ -0,0 +1,282 @@
+    });
+
+    it('should return whether the specified package was local', () => {
+      fs.existsSync.and.returnValues(true, false);
```
nice

petebacondarwin commented on aio/tools/ng-packages-installer.spec.js:
```diff
@@ -0,0 +1,282 @@
+      expect(fs.writeFileSync).toHaveBeenCalledWith(localFlagFile, '');
+    });
+
+    it('should do nothing is the Angular package is not installed', () => {
```
should do nothing **if** the Angular package is not installed

petebacondarwin reviewed (approved):
Just a few minor comments/questions but nothing blocking merging. We will need approval from @IgorMinar for the CI script change.

mary-poppins commented:
You can preview 327f59a at https://pr18722-327f59a.ngbuilds.io/.

gkalpak committed:
refactor(aio): remove unnecessary `--` from yarn commands

gkalpak commented on aio/content/guide/docs-style-guide.md:
```diff
@@ -2,13 +2,13 @@
 
 This page presents design and layout guidelines for Angular documentation pages.  These guidelines should be followed by all guide page authors. Deviations must be approved by the documentation editor.
 
-Most guide pages should have [accompanying sample code](#from-code-samples) with 
```
This file contains a lot of trailing whitespace removals. Better to review with `?w=1`.

gkalpak commented:
Rebased on master, addressed (most of the) comments and added an extra commit to remove unnecessary `--` from yarn commands.

@petebacondarwin, PTAL

mary-poppins commented:
You can preview 6be673d at https://pr18722-6be673d.ngbuilds.io/.

petebacondarwin reviewed (approved):
Still approved :-)

mary-poppins commented:
You can preview 45d133d at https://pr18722-45d133d.ngbuilds.io/.

petebacondarwin commented:
Just waiting on @IgorMinar for approval now

gkalpak commented:
Merged as 7a965dc58 and ffceae0a0.
